### PR TITLE
CORE-6307: Document Corda API: Serialisation

### PR DESCRIPTION
**Pull request by josephzunigadaly**

### CODE REVIEW

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    weight: 8000
+section_menu: corda-5-dev-preview
+---
+
+
+
+# Default Class Evolution
+
+
+Whilst more complex evolutionary modifications to classes require annotating, Corda’s serialization
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:17]: This is hard to follow..
Maybe:

Whilst more complex evolutionary modifications to classes require annotating, Corda’s serialization
framework supports several minor modifications to classes without any external modification to
the actual code changes.

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+For this, we use the annotations to allow developers to express their backward compatible intentions.
+
+In the case of renaming constants this is somewhat obvious, the deserializing node will simply treat any
+constants it doesn’t understand as their “old” values, i.e. those values that it currently knows about.
+
+In the case of adding new constants the developer must chose which constant (that existed *before* adding
+the new one) a deserializing system should treat any instances of the new one as.
+
+{{< note >}}
+Ultimately, this may mean some design compromises are required. If an enumeration is
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:50]: Hard to follow..
Maybe:
Ultimately, this may mean some design compromises are required. If you expect that an enumeration will be extended often and no sensible defaults exist, including a constant
in the original version of the class that all new additions can default to, may make sense.

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+header of the corresponding class. If they match then we are sure that the two class versions are functionally
+the same and no further steps are required save the deserialization of the serialized information into an instance
+of the class.
+
+If, however, the fingerprints differ then we know that the class we are attempting to deserialize is different
+than the version we will be deserializing it into. What we cannot know is which version is newer, at least
+not by examining the fingerprint
+
+{{< note >}}
+Corda’s AMQP fingerprinting for enumerated types include the type name and the enum constants
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:75]:
```suggestion
Corda’s AMQP fingerprinting for enumerated types include the type name and the enum constants.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+{{< /note >}}
+
+Newer vs older is important as the deserializer needs to use the more recent set of transforms to ensure it
+can transform the serialised object into the form as it exists in the deserializer. Newness is determined simply
+by length of the list of all transforms. This is sufficient as transform annotations should only ever be added
+
+
+{{< warning >}}
+technically there is nothing to prevent annotations being removed in newer versions. However,
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:84]:
```suggestion
Technically there is nothing to prevent annotations being removed in newer versions. However,
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+* Determined from the local class and the annotations applied to it (the local copy)
+* Parsed from the AMQP header (the remote copy)
+
+Which set is used will simply be the largest.
+
+## Renaming Constants
+
+Renamed constants are marked as such with the `@CordaSerializationTransformRenames` meta annotation that
+wraps a list of `@CordaSerializationTransformRename` annotations. Each rename requiring an instance in the
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:100]:
```suggestion
wraps a list of `@CordaSerializationTransformRename` annotations. Each rename requires an instance in the
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+)
+enum class Example {
+    A, B, D
+}
+```
+
+{{< note >}}
+The parameters to the `CordaSerializationTransformRename` annotation are defined as ‘to’ and ‘from’,
+so in the above example it can be read as constant D (given that is how the class now exists) was renamed
+from C
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:125]:
```suggestion
from C.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+    CordaSerializationTransformRename(from = "C", to = "D")
+)
+enum class Example {
+    A, E, D
+}
+```
+
+### Rules
+
+* A constant cannot be renamed to match an existing constant, this is enforced through language constraints
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:153]:
```suggestion
* A constant cannot be renamed to match an existing constant. This is enforced through language constraints.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+* A constant cannot be renamed to a value that matches any previous name of any other constant
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:154]:
```suggestion
* A constant cannot be renamed to a value that matches any previous name of any other constant.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+
+If either of these covenants are inadvertently broken, a `NotSerializableException` will be thrown on detection
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:156]:
```suggestion
If either of these covenants are inadvertently broken, a `NotSerializableException` will be thrown
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+by the serialization engine as soon as they are detected. Normally this will be the first time an object doing
+so is serialized. However, in some circumstances, it could be at the point of deserialization.
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:158]: doing what?
Maybe:
Usually, this occurs the first time the object is serialized.

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+)
+enum class Example {
+    A, B, C, D
+}
+```
+
+{{< note >}}
+The parameters to the `CordaSerializationTransformEnumDefault` annotation are defined as ‘newName’ and ‘oldName’,
+so in the above example it can be read as constant D should be treated as constant C if you, the deserializing
+node, don’t know anything about constant D
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:188]:
```suggestion
node, don’t know anything about constant D.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+{{< /note >}}
+
+{{< note >}}
+Just as with the `CordaSerializationTransformRename` transformation if a single transform is being applied
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:192]:
```suggestion
Just as with the `CordaSerializationTransformRename` transformation, if a single transform is being applied,
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+    A, B, C, D
+}
+```
+{{< /tab >}}
+
+{{< /tabs >}}
+
+{{< /note >}}
+New constants may default to any other constant older than them, including constants that have also been added
+since inception. In this example, having added D (above) we add the constant E and chose to default it to D
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:210]:
```suggestion
since inception. In this example, having added D (above) we add the constant E and chose to default it to D.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+    A, B, C, D, E
+}
+```
+{{% /tab %}}
+
+{{< /tabs >}}
+
+{{< note >}}
+Alternatively, we could have decided both new constants should have been defaulted to the first
+element
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md:229]:
```suggestion
element:
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/api-serialisation.md [hunk 15]
```diff
@@ -8,3 +8,618 @@ menu:
...
+instantiate the object with the serialized values.
+
+It is recommended that serializable objects in Corda adhere to the following rules, as they allow immutable state
+objects to be deserialised:
+
+
+
+* A Java Bean getter for each of the properties in the constructor, with a name of the form `getX`.  For example, for a constructor
+  parameter `foo`, there must be a getter called `getFoo()`.  If `foo` is a boolean, the getter may
+  optionally be called `isFoo()` (this is why the class must be compiled with parameter names turned on)
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/api-serialisation.md:286]:
```suggestion
  optionally be called `isFoo()`. This is why the class must be compiled with parameter names turned on.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/api-serialisation.md [hunk 16]
```diff
@@ -8,3 +8,618 @@ menu:
...
+    public void setC(int c) { this.c = c; }
+}
+```
+{{% /tab %}}
+
+{{< /tabs >}}
+
+
+{{< warning >}}
+We do not recommend this pattern! Corda tries to use immutable data structures throughout, and if you
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/api-serialisation.md:370]:
```suggestion
We do not recommend this pattern. Corda tries to use immutable data structures throughout and if you
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/api-serialisation.md [hunk 17]
```diff
@@ -8,3 +8,618 @@ menu:
...
+rely heavily on mutable JavaBean style objects then you may sometimes find the API behaves in unintuitive ways.
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/api-serialisation.md:371]:
```suggestion
rely heavily on mutable JavaBean style objects, you may sometimes find the API behaves in unintuitive ways.
```

**Review by nadinequinr3 — CHANGES_REQUESTED** [review]

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md (hunk 1 above)
Comment (josephzunigadaly) [reply content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:17]: So the suggestion is to replace 'save' with 'to'. My understanding of the sentence is that save means 'other than' or 'except'. I think using the word 'to' changes the meaning but the sentence is still true.

Would you prefer 'to' or 'except'?

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md (hunk 2 above)
Comment (josephzunigadaly) [reply content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md]: Done

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md (hunk 10 above)
Comment (josephzunigadaly) [reply content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md]: Done

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+---
+
+
+
+# Pluggable Serializers for CorDapps
+
+
+To be serializable by Corda, Java classes must be compiled with the -parameters switch to enable matching of its properties
+to constructor parameters. This is important because Corda’s internal AMQP serialization scheme will only construct
+objects using their constructors. However, when recompilation isn’t possible, or classes are built in such a way that
```
[the remark this answers is not in the document]
Comment (josephzunigadaly) [reply content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md]: Done

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md (hunk 1 above)
Comment (nadinequinr3) [reply content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:17]: "except" please. I had written it with that initially but was unsure if I'd missed understood! Thanks

Comment (josephzunigadaly) [reply content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:17]: Done

**Review by nadinequinr3 — APPROVED** [review]

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+
+
+# Pluggable Serializers for CorDapps
+
+
+To be serializable by Corda, Java classes must be compiled with the -parameters switch to enable matching of its properties
+to constructor parameters. This is important because Corda’s internal AMQP serialization scheme will only construct
+objects using their constructors. However, when recompilation is not possible, or classes are built in such a way that
+they cannot be easily modified for simple serialization, CorDapps can provide custom proxy serializers. Corda
+can use these serializers to move from types it cannot serialize to an interim representation that it can serialize.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:21]:
```suggestion
can use these serializers to move from a type it cannot serialize to an interim representation that it can serialize.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+The transformation to and from this proxy object are handled by the supplied serializer.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:22]:
```suggestion
The transformation to and from this proxy object is handled by the supplied serializer.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+  corda-5-dev-preview:
+    identifier: corda-5-dev-preview-api-serialisation-cordapp-custom-serializers
+    parent: corda-5-dev-preview-api-serialisation
+    weight: 8000
+section_menu: corda-5-dev-preview
+---
+
+
+
+# Pluggable Serializers for CorDapps
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:14]:
```suggestion
# Pluggable serializers for CorDapps
```

Sentence case rather than the usual title case is used for headings.

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+to constructor parameters. This is important because Corda’s internal AMQP serialization scheme will only construct
+objects using their constructors. However, when recompilation is not possible, or classes are built in such a way that
+they cannot be easily modified for simple serialization, CorDapps can provide custom proxy serializers. Corda
+can use these serializers to move from types it cannot serialize to an interim representation that it can serialize.
+The transformation to and from this proxy object are handled by the supplied serializer.
+
+
+## Writing a Custom Serializer
+
+Serializers must
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:27]:
```suggestion
Serializers must:
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+
+
+
+* Inherit from `net.corda.v5.serialization.SerializationCustomSerializer`
+* Provide a proxy class to transform the object to and from
+* Implement the `toProxy` and `fromProxy` methods
+* Be included into the CorDapp package
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:34]:
```suggestion
* Be included in the CorDapp package
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    public static Example of (int[] a) { return Example(a[0], a[1]); }
+
+    public int getA() { return a; }
+    public int getB() { return b; }
+}
+```
+
+Without a custom serializer we cannot serialize this class as there is no public constructor that facilitates the
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:64]:
```suggestion
Without a custom serializer, we cannot serialize this class as there is no public constructor that facilitates the
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+        return Example.of(constructorArg)
+    }
+}
+```
+{{% /tab %}}
+
+{{< /tabs >}}
+
+In the above examples
+
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:191]: Delete one of the two blank lines

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+
+* `ExampleSerializer` is the actual serializer that will be loaded by the framework to serialize instances of the `Example` type.
+* `ExampleSerializer.Proxy`, in the Kotlin example, and `ExampleProxy` in the Java example, is the intermediate representation used by the framework to represent instances of `Example` within the wire format.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:194]:
```suggestion
* `ExampleSerializer.Proxy`, in the Kotlin example, and `ExampleProxy`, in the Java example, is the intermediate representation used by the framework to represent instances of `Example` within the wire format.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+
+
+## The Proxy Object
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:197]:
```suggestion
## The proxy object
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 28]
```diff
@@ -0,0 +1,244 @@
...
+
+The proxy object should be thought of as an intermediate representation that the serialization framework
+can reason about. One is being written for a class because, for some reason, that class cannot be
+introspected successfully but that framework. It is therefore important to note that the proxy class must
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:201]:
```suggestion
introspected successfully but that framework. 
```

Unclear meaning; is there something accidentally missing?

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 29]
```diff
@@ -0,0 +1,244 @@
...
+only contain elements that the framework can reason about.
+
+The proxy class itself is distinct from the proxy serializer. The serializer must refer to the unserializable
+type in the `toProxy` and `fromProxy` methods.
+
+For example, the first thought a developer may have when implementing a proxy class is to simply *wrap* an
+instance of the object being proxied. This is shown below
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:208]:
```suggestion
instance of the object being proxied. This is shown below:
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 30]
```diff
@@ -0,0 +1,244 @@
...
+
+    override fun fromProxy(proxy: Proxy) : Example {
+        return proxy.e
+    }
+}
+```
+
+However, this will not work because what we’ve created is a recursive loop whereby synthesising a serializer
+for the `Example` type requires synthesising one for `ExampleSerializer.Proxy`. However, that requires
+one for `Example` and so on and so forth until we get a `StackOverflowException`.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:227]:
```suggestion
one for `Example` and so on until we get a `StackOverflowException`.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 31]
```diff
@@ -0,0 +1,244 @@
...
+
+The solution, as shown initially, is to create the intermediate form (the Proxy object) purely in terms
+the serialization framework can reason about.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:230]:
```suggestion
which the serialization framework can reason about.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 32]
```diff
@@ -0,0 +1,244 @@
...
+
+
+{{< important >}}
+When composing a proxy object for a class be aware that everything within that structure will be written
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:234]:
```suggestion
When composing a proxy object for a class, be aware that everything within that structure will be written
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 33]
```diff
@@ -0,0 +1,244 @@
...
+into the serialized byte stream.
+
+
+{{< /important >}}
+
+
+## Allow list
+
+By writing a custom serializer for a class it has the effect of adding that class to the allow list, meaning such
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:243]:
```suggestion
By writing a custom serializer for a class, it has the effect of adding that class to the allow list, meaning such
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md [hunk 34]
```diff
@@ -0,0 +1,244 @@
...
+classes don’t need the `@CordaSerializable` annotation.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-cordapp-custom-serializers.md:244]:
```suggestion
classes do not need the `@CordaSerializable` annotation.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 35]
```diff
@@ -0,0 +1,287 @@
...
+  corda-5-dev-preview:
+    identifier: corda-5-dev-preview-api-serialisation-default-evolution
+    parent: corda-5-dev-preview-api-serialisation
+    weight: 8000
+section_menu: corda-5-dev-preview
+---
+
+
+
+# Default Class Evolution
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:14]:
```suggestion
# Default class evolution
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 36]
```diff
@@ -0,0 +1,287 @@
...
+data class Example1 (val a: Int, b: String) // (Version A)
+
+// Class post addition of property c
+data class Example1 (val a: Int, b: String, c: Int?) // (Version B)
+```
+{{% /tab %}}
+
+{{< /tabs >}}
+
+A node with version A of class `Example1`  will be able to deserialize a blob serialized by a node with it
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:47]:
```suggestion
A node with version A of class `Example1` will be able to deserialize a blob serialized by a node with it
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 37]
```diff
@@ -0,0 +1,287 @@
...
+at version B as the framework would treat it as a removed property.
+
+A node with the class at version B will be able to deserialize a serialized version A of `Example1` without
+any modification as the property is nullable and will thus provide null to the constructor.
+
+
+## Adding Non Nullable Properties
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:54]:
```suggestion
## Adding non-nullable properties
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 38]
```diff
@@ -0,0 +1,287 @@
...
+
+If a non null property is added, unlike nullable properties, some additional code is required for
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:56]:
```suggestion
If a non-null property is added, unlike nullable properties, some additional code is required for
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 39]
```diff
@@ -0,0 +1,287 @@
...
+this to work. Consider a similar example to our nullable example above
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:57]:
```suggestion
this to work. Consider a similar example to our nullable example above:
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 40]
```diff
@@ -0,0 +1,287 @@
...
+data class Example1 (val a: Int, b: String, c: Int) { // (Version B)
+     @DeprecatedConstructorForDeserialization(1)
+     constructor (a: Int, b: String) : this(a, b, 0) // 0 has been determined as a sensible default
+}
+```
+{{% /tab %}}
+
+{{< /tabs >}}
+
+For this to work we have had to add a new constructor that allows nodes with the class at version B to create an
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:75]:
```suggestion
For this to work, we would have to add a new constructor that allows nodes with the class at version B to create an
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 41]
```diff
@@ -0,0 +1,287 @@
...
+instance from serialised form of that class from an older version, in this case version A as per our example
+above. A sensible default for the missing value is provided for instantiation of the non null property.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:77]:
```suggestion
above. A sensible default for the missing value is provided for instantiation of the non-null property.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 42]
```diff
@@ -0,0 +1,287 @@
...
+
+{{< note >}}
+The `@DeprecatedConstructorForDeserialization` annotation is important, this signifies to the
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:80]:
```suggestion
The `@DeprecatedConstructorForDeserialization` annotation is important; this signifies to the
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 43]
```diff
@@ -0,0 +1,287 @@
...
+serialization framework that this constructor should be considered for building instances of the
+object when evolution is required.
+
+Furthermore, the integer parameter passed to the constructor if the annotation indicates a precedence
+order, see the discussion below.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:85]:
```suggestion
order; see the discussion below.
```

File: content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md [hunk 44]
```diff
@@ -0,0 +1,287 @@
...
+
+{{< /note >}}
+As before, instances of the class at version A will be able to deserialize serialized forms of example B as it
+will simply treat them as if the property has been removed (as from its perspective, they will have been).
+
+
+### Constructor Versioning
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-default-evolution.md:92]:
```suggestion
### Constructor versioning
```

**Comment by nadinequinr3** [comment]
Merging to make available for QA
